### PR TITLE
[cuda] Add built-in UVM allocator context manager for numerics testing

### PR DESCRIPTION
**Pull request by IvanKobzarev**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.15.0) (oldest at bottom):
* __->__ #185411

Add `torch.cuda._use_uvm()`, a context manager that routes CUDA
allocations through `cudaMallocManaged` (Unified Virtual Memory). UVM
transparently pages between GPU device memory and host RAM, allowing a
single GPU to run workloads that exceed its physical VRAM. Numerics are
identical to regular device allocations.

The primary use case is numerics validation of precompiled graph trainer
artifacts. When testing precompile artifacts from large multi-node jobs
(e.g., Llama3-70B on 64 GPUs with TP=8 DP=8), we run the artifact
single-process with FakePG simulating the full world. The traced FX
graph materializes all layers' FSDP-allgathered weights at once (no
per-layer resharding in the flat graph), peaking at ~141 GB -- well
beyond a single H100's 95 GB.

To verify correctness, we compare the precompile artifact against the
eager DTensor reference path. The eager path can consume even more
memory than precompile (101+ GB in practice) because it materializes
full allgathered weights during forward and retains activations for
backward. Without UVM, the eager reference OOMs on a single GPU, making
the comparison impossible. With UVM, both paths run to completion and we
can verify bitwise-identical loss and gradient hashes, confirming the
precompile artifact is numerically faithful.

This is a correctness tool, not a performance tool -- UVM page migration
adds 5-10x overhead. The allocator is thread-local: only the calling
thread's allocations go through managed memory. For eager models where
autograd backward runs on a separate C++ thread, pair with
`torch.autograd.set_multithreading_enabled(False)` so backward
allocations also use UVM.

The implementation adds a pybind binding that creates a
CUDAPluggableAllocator with cudaMallocManaged/cudaFree lambdas and
cudaMemAdvise hints (SetPreferredLocation + SetAccessedBy). The Python
context manager wraps this in a MemPool via the existing use_mem_pool
infrastructure. No new C++ classes or files needed.

Authored by Claude.

Test Plan:

Added three unit tests in TestMemPool (test/test_cuda.py):
- test_use_uvm: allocate + matmul under UVM, verify result
- test_use_uvm_numerics: same seed produces identical tensors in UVM vs regular
- test_use_uvm_backward: forward + backward through nn.Linear with
  single-threaded autograd

```
python test/test_cuda.py -k test_use_uvm
```

Validated on devgpu (H100 95GB) with Llama3-70B precompile artifact
(141.5 GB peak, ~46 GB paged to host) and eager DTensor reference
(101+ GB peak). Both paths produced bitwise-identical loss and gradient
hashes across all training steps.

### CODE REVIEW

**Review by ngimel — APPROVED** [review]

File: torch/cuda/memory.py [hunk 1]
```diff
@@ -1357,3 +1357,100 @@ def use_mem_pool(pool: MemPool, device: "Device" = None):
...
+                )
+            )
+            _check_err(
+                cuda_runtime.cudaMemAdvise(
+                    ptr,
+                    size,
+                    cuda_runtime.cudaMemoryAdvise.cudaMemAdviseSetAccessedBy,
+                    location,
+                )
+            )
```
Comment (ngimel) [inline torch/cuda/memory.py:1473]: try doing better error handling - I think error will be just ignored, and you'll return some garbage ptr. It'll make sense to return nullptr then CUDACachingAllocator can take over and throw an error (even though it'll be confusing), see e.g. https://www.internalfb.com/code/fbsource/[af3c68fb571abb58ff811ec3ef19fbc5abb33618]/genai/msl/ops/kernels/cuda/vmm_pool.py?lines=433

Comment (IvanKobzarev) [reply torch/cuda/memory.py:1473]: Thanks, I will try to improve it and test error handling/propagation.

**Comment by huydhn** [comment]
@pytorchbot revert -m 'Sorry for reverting your change but trunk tests were not run in the PR due to an infra sev yesterday and the test is failing in trunk' -c weird

test/test_cuda_expandable_segments.py::TestMemPool::test_use_uvm_numerics [GH job link](https://github.com/pytorch/pytorch/actions/runs/26662106904/job/78607262432) [HUD commit link](https://hud.pytorch.org/pytorch/pytorch/commit/9a413a79d49e8a921360d5e33918017ead1c28b0)
